scanner/src/main/kotlin/experimental/ExperimentalScanner.kt

```diff
@@ -213,33 +214,69 @@ class ExperimentalScanner(
      * Run package scanners for packages with incomplete scan results.
      */
     private fun runPackageScanners(controller: ScanController, context: ScanContext) {
-        controller.packages.forEach { pkg ->
-            val nestedProvenance = controller.getNestedProvenance(pkg.id) ?: return@forEach
-
-            // TODO: Use coroutines to execute scanners in parallel.
+        controller.getPackagesConsolidatedByProvenance().forEach { (provenance, packages) ->
             controller.getPackageScanners().forEach scanner@{ scanner ->
-                if (controller.hasCompleteScanResult(scanner, pkg)) {
-                    log.debug {
-                        "Skipping scan of '${pkg.id.toCoordinates()}' with package scanner '${scanner.name}' as " +
-                                "stored results are available."
+                val packagesWithIncompleteScanResult = packages
+                    .filter { pkg ->
+                        (controller.getNestedProvenance(pkg.id) != null).also { hasNestedProvenance ->
+                            if (!hasNestedProvenance) {
+                                log.debug {
+                                    "Skipping scan of '${pkg.id.toCoordinates()}' with package scanner " +
+                                            "${scanner.name} because not nested provenance for the package could be " +
+                                            "resolved."
+                                }
+                            }
+                        }
+                    }
+                    .filter { pkg ->
+                        !controller.hasCompleteScanResult(scanner, pkg).also { hasCompleteScanResult ->
+                            if (!hasCompleteScanResult) {
+                                log.debug {
+                                    "Skipping scan of '${pkg.id.toCoordinates()}' with package scanner " +
+                                            "'${scanner.name}' as stored results are available."
+                                }
+                            }
+                        }
                     }
 
-                    return@scanner
+                if (packagesWithIncompleteScanResult.isEmpty()) return@scanner
+
+                // Create a reference package with any VCS path removed, to ensure the full repository is scanned.
```

Comment:
Is this block still necessary, as `getPackagesConsolidatedByProvenance()` should already do that, or?

Reply:
The function does only group the packages, it does not create a reference package.

Reply:
Oh right, the function removes the VCS path from the provenance, not from the package.

Reply:
Yes, the packages need to keep the path to correctly filter the scan results later on.